server/resp/src/test/java/org/infinispan/server/resp/JsonCommandsTest.java

```diff
@@ -0,0 +1,603 @@
+      result = redis.jsonGet(key, jpRoot);
+      assertThat(result).hasSize(1);
+      assertThat(
+            compareJSONSet(result.get(0).asJsonArray().getFirst(), jv0.asJsonArray().getFirst(), "$.key1b.key2b", jv))
+            .isEqualTo(true);
+   }
+
+   @Test
+   public void testJSONSETLegacy() {
+      CustomStringCommands command = CustomStringCommands.instance(redisConnection);
+      String key = k();
+      String doc = "{\"key\":\"value\"}";
+      var jsonDoc = defaultJsonParser.createJsonValue(doc);
+      String p = ".";
+      var jp = new JsonPath(p);
+      // Test create root object
+      assertThat(command.jsonSet(key, ".", doc)).isEqualTo("OK");
+      var result = defaultJsonParser.createJsonValue(command.jsonGet(key, "."));
+      assertThat(compareJSONGet(result, jsonDoc, jp)).isEqualTo(true);
+      // Test root can be updated
+      doc = """
+            {
+            "key": { "key1": "val1" }
+            }
+            """;
+      assertThat(command.jsonSet(key, ".", doc)).isEqualTo("OK");
+      result = defaultJsonParser.createJsonValue(command.jsonGet(key, "."));
+      jsonDoc = defaultJsonParser.createJsonValue(doc);
+      assertThat(compareJSONGet(result, jsonDoc, new JsonPath("."))).isEqualTo(true);
+      // Test adding a field in a leaf
+      String v1 = "{\"key2\":\"value2\"}";
+      assertThat(command.jsonSet(key, ".key.key", v1)).isEqualTo("OK");
+      result = defaultJsonParser.createJsonValue(command.jsonGet(key, "."));
+      var jv1 = defaultJsonParser.createJsonValue(v1);
+      assertThat(compareJSONSet(result, jsonDoc, "$.key.key", jv1)).isEqualTo(true);
+   }
+
+   public void testJSONSETWithPath() {
+      String key = k();
+      // Create json
+      JsonPath jpRoot = new JsonPath("$");
+      JsonValue jvDoc = defaultJsonParser.createJsonValue("{\"root\": { \"k1\" : \"v1\", \"k2\":\"v2\"}}");
+      assertThat(redis.jsonSet(key, jpRoot, jvDoc)).isEqualTo("OK");
+      // Modify json
+      JsonPath jpLeaf = new JsonPath("$.root.k1");
+      JsonValue jvNew = defaultJsonParser.createJsonValue("\"newv1\"");
+      assertThat(redis.jsonSet(key, jpLeaf, jvNew)).isEqualTo("OK");
+      // Verify
+      var result = redis.jsonGet(key, jpRoot);
+      assertThat(result).hasSize(1);
+      assertThat(compareJSONSet(result.get(0).asJsonArray().getFirst(), jvDoc, "$.root.k1", jvNew));
+   }
+
+   public void testJSONSETInvalidPath() {
+      String key = k();
+      // Create json
+      JsonPath jpRoot = new JsonPath("$");
+      JsonValue jvDoc = defaultJsonParser.createJsonValue("{\"root\": { \"k1\" : \"v1\", \"k2\":\"v2\"}}");
+      assertThat(redis.jsonSet(key, jpRoot, jvDoc)).isEqualTo("OK");
+      CustomStringCommands command = CustomStringCommands.instance(redisConnection);
+      // Modify json
+      assertThatThrownBy(() -> {
+         command.jsonSet(key, "", "\"newValue\"");
+      }).isInstanceOf(RedisCommandExecutionException.class);
```

Comment:
If you try invalid JSON with Redis CLI it'll crash. Some quick internet search show that Redis CLI might not be handling well new lines if sent with the response. (But I don't see that we'd be sending a new lines with errors anywhere)

```
$ redis-cli
127.0.0.1:6379> JSON.SET json $ {not-a-json
Error: Bad simple string value
$ ....
```

Reply:
Yep `\n` is not good. It happens that jsonpath exception returns message with newline and it hangs the server. This exception also contains info about the error in the json value so it's nice for the user to have it back. I've escaped the `\n` as suggested by @jabolina in the error serialization and the redis testsuite looks good also with that

Reply:
Can you give it a shot with redis-cli? It seems that the issue still persist there

Reply:
yeah the escaping was wrong. there's no way to pass 0xa char into a simple string, replaced with "\n"